### PR TITLE
project#self.valid_name?

### PR DESCRIPTION
**Pull request by Ana06**
Migrate `project#self.valid_name? `to the new test suite and refactor `project#self.valid_name?` :bowtie:

### CODE REVIEW

File: src/api/app/models/project.rb [hunk 1]
```diff
@@ -1404,13 +1404,10 @@ def bsrequest_repos_map(project)
 
   def self.valid_name?(name)
```
Comment (ChrisBr) [inline src/api/app/models/project.rb]: Where do we use ``valid_name?``? Only as validation? Maybe we can try to move the functionality in plain ActiveModel validations like ``format_of`` and ``length`` etc?

File: src/api/spec/models/project_spec.rb [hunk 2]
```diff
@@ -404,4 +405,31 @@
     it { expect(leap_project.image_template?).to be(true) }
     it { expect(tumbleweed_project.image_template?).to be(false) }
   end
+
+  describe '#self.valid_name?' do
```
Comment (ChrisBr) [inline src/api/spec/models/project_spec.rb]: What about using shoulda matchers here?

File: src/api/app/models/project.rb (hunk 1 above)
Comment (Ana06) [inline src/api/app/models/project.rb]: The problem of using `format_of` is that we should write a regular expression that match, and here for some conditions I think it is match clear to write a regular expression that doesn't match, like we do with `return false if name =~ /::/`

And it is not only used as validation, for example:

- https://github.com/openSUSE/open-build-service/blob/2889f4d98df379f29bfe3cc47ec9eadf0eff259b/src/api/app/helpers/validation_helper.rb#L11

- https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/branch_package.rb#L589

- https://github.com/openSUSE/open-build-service/blob/3271934e216593ddaf2f3576a125abdb3d707d6c/src/api/app/jobs/consistency_check.rb#L18

File: src/api/spec/models/project_spec.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/spec/models/project_spec.rb]: Hm, if we leave it as a helper function not possible ... Nevermind!

File: src/api/spec/models/project_spec.rb [hunk 3]
```diff
@@ -404,4 +404,30 @@
...
+      # has ::
+      it{ expect(Project.valid_name?('home:M0ses:raspi::qtdesktop')).to be(false) }
+      # 0
+      it{ expect(Project.valid_name?('0')).to be(false) }
+      # end with :
+      it{ expect(Project.valid_name?('home:M0ses:')).to be(false) }
+      # starts with _
+      it{ expect(Project.valid_name?('_foobar')).to be(false) }
+      # More than 200 characters
+      it{ expect(Project.valid_name?("4" * 250)).to be(false) }
```
Comment (ChrisBr) [inline src/api/spec/models/project_spec.rb]: Test the corner case! 201 character!

File: src/api/spec/models/project_spec.rb [hunk 4]
```diff
@@ -404,4 +404,30 @@
...
+      # blank
+      it{ expect(Project.valid_name?('')).to be(false) }
+    end
+
+    context "valid" do
+      it{ expect(Project.valid_name?("foobar")).to be(true) }
```
Comment (ChrisBr) [inline src/api/spec/models/project_spec.rb]: What about
```
it{ expect(Project.valid_name?('home:M0ses:raspi:qtdesktop')).to be(true) }
```

File: src/api/app/models/project.rb [hunk 5]
```diff
@@ -1404,13 +1404,10 @@ def bsrequest_repos_map(project)
...
     return false unless name.kind_of? String
-    # this length check is duplicated but useful for other uses for this function
-    return false if name.length > 200 || name.blank?
     return false if name == "0";
-    return false if name =~ %r{^[_\.]}
-    return false if name =~ %r{::}
+    return false if name =~ /::/
     return false if name.end_with?(':')
-    return true if name =~ /\A\w[-+\w\.:]*\z/
+    return true if name =~ /\A[a-zA-Z0-9][-+\w\.:]{0,199}\z/
```
Comment (ChrisBr) [inline src/api/app/models/project.rb]: Hm, I think ``name.length > 200`` is more explicit and probably even faster ... But I'm fine with it if you like it more!

File: src/api/spec/models/project_spec.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/spec/models/project_spec.rb]: Btw: Do we really need the comments? I don't like it ...

Comment (Ana06) [inline src/api/spec/models/project_spec.rb]: I can remove then

File: src/api/spec/models/project_spec.rb (hunk 3 above)
Comment (Ana06) [inline src/api/spec/models/project_spec.rb]: Agree! It was already like that in the old test suite!

File: src/api/app/models/project.rb (hunk 5 above)
Comment (Ana06) [inline src/api/app/models/project.rb]: but as it was said in the comment the length checked is already duplicated, so not that important to make it explicit. And if you want to know what the function does you have to find out what the regular expression does so I think that there is not much difference.
If it is longer it may be faster, but if not I would say it the same as you have to check all the character in the String anyway.

Comment (bgeuken) [inline src/api/app/models/project.rb]: @Ana06 This seems to make the validation more strict. In that case, please make sure that this won't invalidate any existing projects.
Or at least that such projects still could be fixed by users. Other option would be to provide a migration replacing invalid chars.

Comment (Ana06) [inline src/api/app/models/project.rb]: @bgeuken 

> This seems to make the validation more strict

It is the same, I just thought that it was more clear to write it like that instead of having several conditions. :wink:
